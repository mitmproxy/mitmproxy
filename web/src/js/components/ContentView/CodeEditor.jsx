import React, { Component, PropTypes } from 'react'
import { render } from 'react-dom';
import Codemirror from 'react-codemirror';


CodeEditor.propTypes = {
        content: PropTypes.string.isRequired,
        onChange: PropTypes.func.isRequired,
}

export default function CodeEditor ( { content, onChange} ){

    let options = {
        lineNumbers: true
    };
    return (
        <div className="codeeditor" onKeyDown={e => e.stopPropagation()}>
            <Codemirror value={content} onChange={onChange} options={options}/>
        </div>
    )
}
