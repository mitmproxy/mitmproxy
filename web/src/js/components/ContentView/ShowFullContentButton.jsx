import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import { render } from 'react-dom';
import Button from '../common/Button';
import { setShowFullContent } from '../../ducks/ui/flow'



ShowFullContentButton.propTypes = {
        setShowFullContent: PropTypes.func.isRequired,
        showFullContent: PropTypes.bool.isRequired
}

function ShowFullContentButton ( {setShowFullContent, showFullContent, visibleLines, contentLines} ){

    return (
        !showFullContent &&
            <div>
                <Button className="view-all-content-btn btn-xs" onClick={() => setShowFullContent()} text="Show full content"/>
                <span className="pull-right"> {visibleLines}/{contentLines} are visible &nbsp; </span>
            </div>
    )
}

export default connect(
    state => ({
        showFullContent: state.ui.flow.showFullContent,
        visibleLines: state.ui.flow.maxContentLines,
        contentLines: state.ui.flow.content.length

    }),
    {
        setShowFullContent
    }
)(ShowFullContentButton)

