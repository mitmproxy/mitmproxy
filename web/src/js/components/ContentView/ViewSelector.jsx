import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { setContentView } from '../../ducks/ui/flow';
import Dropdown from '../common/Dropdown'


ViewSelector.propTypes = {
    contentViews: PropTypes.array.isRequired,
    activeView: PropTypes.string.isRequired,
    setContentView: PropTypes.func.isRequired
}

export function ViewSelector ({contentViews, activeView, setContentView}){
    let inner = <span> <b>View:</b> {activeView.toLowerCase()} <span className="caret"/> </span>

    return (
        <Dropdown dropup className="pull-left" btnClass="btn btn-default btn-xs" text={inner}>
            {contentViews.map(name =>
                <a href="#" key={name}  onClick={e => {e.preventDefault(); setContentView(name)}}>
                    {name.toLowerCase().replace('_', ' ')}
                </a>
                )
            }
        </Dropdown>
    )
}

export default connect (
    state => ({
        contentViews: state.settings.contentViews || [],
        activeView: state.ui.flow.contentView,
    }), {
        setContentView,
    }
)(ViewSelector)
