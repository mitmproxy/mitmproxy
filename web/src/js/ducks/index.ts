import {
    AnyAction,
    applyMiddleware,
    combineReducers,
    compose,
    createStore as createReduxStore,
    PreloadedState,
    Store
} from "redux"
import eventLog from "./eventLog"
import flows from "./flows"
import ui from "./ui/index"
import connection from "./connection"
import options from './options'
import commandBar from "./commandBar";
import thunk, {ThunkAction, ThunkDispatch, ThunkMiddleware} from "redux-thunk";
import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import backendState from "./backendState";
import options_meta from "./options_meta";


// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const rootReducer = combineReducers({
    commandBar,
    eventLog,
    flows,
    connection,
    ui,
    options,
    options_meta,
    backendState,
});
export type RootState = ReturnType<typeof rootReducer>


export type AppThunk<ReturnType = void> = ThunkAction<ReturnType,
    RootState,
    unknown,
    AnyAction>

export const createAppStore = (preloadedState?: PreloadedState<RootState>): Store<RootState> => {
    return createReduxStore(
        rootReducer,
        preloadedState,
        composeEnhancers(applyMiddleware(
            thunk as ThunkMiddleware<RootState>
        )))
};

export const store = createAppStore(undefined);

// this would be correct, but PyCharm bails on it
// export type AppDispatch = typeof store.dispatch
// instead:
export type AppDispatch = ThunkDispatch<RootState, void, AnyAction>
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
