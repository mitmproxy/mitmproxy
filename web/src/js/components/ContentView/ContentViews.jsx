import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { setContentViewDescription, setContent } from '../../ducks/ui/flow'
import withContentLoader from './ContentLoader'
import { MessageUtils } from '../../flow/utils'
import CodeEditor from './CodeEditor'


const isImage = /^image\/(png|jpe?g|gif|webp|vnc.microsoft.icon|x-icon)$/i
ViewImage.matches = msg => isImage.test(MessageUtils.getContentType(msg))
ViewImage.propTypes = {
    flow: PropTypes.object.isRequired,
    message: PropTypes.object.isRequired,
}
function ViewImage({ flow, message }) {
    return (
        <div className="flowview-image">
            <img src={MessageUtils.getContentURL(flow, message)} alt="preview" className="img-thumbnail"/>
        </div>
    )
}

Edit.propTypes = {
    content: PropTypes.string.isRequired,
}

function Edit({ content, onChange }) {
    return <CodeEditor content={content} onChange={onChange}/>
}
Edit = withContentLoader(Edit)

export class PureViewServer extends Component {
    static propTypes  = {
        showFullContent: PropTypes.bool.isRequired,
        maxLines: PropTypes.number.isRequired,
        setContentViewDescription : PropTypes.func.isRequired,
        setContent: PropTypes.func.isRequired
    }

    componentWillMount(){
        this.setContentView(this.props)
    }

    componentWillReceiveProps(nextProps){
        if (nextProps.content != this.props.content) {
            this.setContentView(nextProps)
        }
    }

    setContentView(props){
        try {
            this.data = JSON.parse(props.content)
        }catch(err) {
            this.data = {lines: [], description: err.message}
        }

        props.setContentViewDescription(props.contentView != this.data.description ? this.data.description : '')
        props.setContent(this.data.lines)
    }

    render() {
        const {content, contentView, message, maxLines} = this.props
        let lines = this.props.showFullContent ? this.data.lines : this.data.lines.slice(0, maxLines)
        return (
            <div>
                {ViewImage.matches(message) && <ViewImage {...this.props} />}
                <pre>
                    {lines.map((line, i) =>
                        <div key={`line${i}`}>
                            {line.map((element, j) => {
                                let [style, text] = element
                                return (
                                    <span key={`tuple${j}`} className={style}>
                                        {text}
                                    </span>
                                )
                            })}
                        </div>
                    )}
                </pre>
            </div>
        )
    }

}

const ViewServer = connect(
    state => ({
        showFullContent: state.ui.flow.showFullContent,
        maxLines: state.ui.flow.maxContentLines
    }),
    {
        setContentViewDescription,
        setContent
    }
)(withContentLoader(PureViewServer))

export { Edit, ViewServer, ViewImage }
