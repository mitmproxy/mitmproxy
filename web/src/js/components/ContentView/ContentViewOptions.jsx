import React, { PropTypes } from 'react'
import { connect } from 'react-redux'
import ViewSelector from './ViewSelector'
import UploadContentButton from './UploadContentButton'
import DownloadContentButton from './DownloadContentButton'

ContentViewOptions.propTypes = {
    flow: React.PropTypes.object.isRequired,
    message: React.PropTypes.object.isRequired,
}

function ContentViewOptions(props) {
    const { flow, message, uploadContent, readonly, contentViewDescription } = props
    return (
        <div className="view-options">
            <ViewSelector message={message}/>
            &nbsp;
            <DownloadContentButton flow={flow} message={message}/>
            &nbsp;
            <UploadContentButton uploadContent={uploadContent}/>
            &nbsp;
            <span>{contentViewDescription}</span>
        </div>
    )
}

export default connect(
    state => ({
        contentViewDescription: state.ui.flow.viewDescription
    })
)(ContentViewOptions)
