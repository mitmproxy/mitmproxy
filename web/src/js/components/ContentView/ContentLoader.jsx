import React, { Component, PropTypes } from 'react'
import { MessageUtils } from '../../flow/utils.js'

export default View => class extends React.Component {

    static displayName = View.displayName || View.name
    static matches = View.matches

    static propTypes = {
        ...View.propTypes,
        content: PropTypes.string,  // mark as non-required
        flow: PropTypes.object.isRequired,
        message: PropTypes.object.isRequired,
    }

    constructor(props) {
        super(props)
        this.state = {
            content: undefined,
            request: undefined,
        }
    }

    componentWillMount() {
        this.updateContent(this.props)
    }

    componentWillReceiveProps(nextProps) {
        if (
            nextProps.message.content !== this.props.message.content ||
            nextProps.message.contentHash !== this.props.message.contentHash ||
            nextProps.contentView !== this.props.contentView
        ) {
            this.updateContent(nextProps)
        }
    }

    componentWillUnmount() {
        if (this.state.request) {
            this.state.request.abort()
        }
    }

    updateContent(props) {
        if (this.state.request) {
            this.state.request.abort()
        }
        // We have a few special cases where we do not need to make an HTTP request.
        if(props.message.content !== undefined) {
            return this.setState({request: undefined, content: props.message.content})
        }
        if(props.message.contentLength === 0 || props.message.contentLength === null){
            return this.setState({request: undefined, content: ""})
        }

        let requestUrl = MessageUtils.getContentURL(props.flow, props.message, (View.name == 'ViewServer' ? props.contentView : undefined))

        // We use XMLHttpRequest instead of fetch() because fetch() is not (yet) abortable.
        let request = new XMLHttpRequest();
        request.addEventListener("load", this.requestComplete.bind(this, request));
        request.addEventListener("error", this.requestFailed.bind(this, request));
        request.open("GET", requestUrl);
        request.send();
        this.setState({ request, content: undefined })
    }

    requestComplete(request, e) {
        if (request !== this.state.request) {
            return // Stale request
        }
        this.setState({
            content: request.responseText,
            request: undefined
        })
    }

    requestFailed(request, e) {
        if (request !== this.state.request) {
            return // Stale request
        }
        console.error(e)
        // FIXME: Better error handling
        this.setState({
            content: "Error getting content.",
            request: undefined
        })
    }

    render() {
        return this.state.content !== undefined ? (
            <View content={this.state.content} {...this.props}/>
        ) : (
            <div className="text-center">
                <i className="fa fa-spinner fa-spin"></i>
            </div>
        )
    }
};
