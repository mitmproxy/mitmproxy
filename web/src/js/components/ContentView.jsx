import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import * as ContentViews from './ContentView/ContentViews'
import * as MetaViews from './ContentView/MetaViews'
import ShowFullContentButton from './ContentView/ShowFullContentButton'


import { displayLarge, updateEdit } from '../ducks/ui/flow'

ContentView.propTypes = {
    // It may seem a bit weird at the first glance:
    // Every view takes the flow and the message as props, e.g.
    // <Auto flow={flow} message={flow.request}/>
    flow: React.PropTypes.object.isRequired,
    message: React.PropTypes.object.isRequired,
}

ContentView.isContentTooLarge = msg => msg.contentLength > 1024 * 1024 * (ContentViews.ViewImage.matches(msg) ? 10 : 0.2)

function ContentView(props) {
    const { flow, message, contentView, isDisplayLarge, displayLarge, onContentChange, readonly } = props

    if (message.contentLength === 0 && readonly) {
        return <MetaViews.ContentEmpty {...props}/>
    }

    if (message.contentLength === null && readonly) {
        return <MetaViews.ContentMissing {...props}/>
    }

    if (!isDisplayLarge && ContentView.isContentTooLarge(message)) {
        return <MetaViews.ContentTooLarge {...props} onClick={displayLarge}/>
    }

    const View = ContentViews[contentView] || ContentViews['ViewServer']
    return (
        <div className="contentview">
            <View flow={flow} message={message} contentView={contentView} readonly={readonly} onChange={onContentChange}/>
            <ShowFullContentButton/>
        </div>
    )
}

export default connect(
    state => ({
        contentView: state.ui.flow.contentView,
        isDisplayLarge: state.ui.flow.displayLarge,
    }),
    {
        displayLarge,
        updateEdit
    }
)(ContentView)
