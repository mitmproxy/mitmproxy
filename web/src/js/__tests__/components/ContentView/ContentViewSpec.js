import React from 'react'
import renderer from 'react-test-renderer'
import TestUtils from 'react-dom/test-utils'
import { Provider } from 'react-redux'
import { ViewServer, ViewImage, PureViewServer, Edit } from '../../../components/ContentView/ContentViews'
import { TFlow, TStore } from '../../ducks/tutils'
import mockXMLHttpRequest from 'mock-xmlhttprequest'

window.XMLHttpRequest = mockXMLHttpRequest
let tflow = new TFlow()

describe('ViewImage Component', () => {
    let viewImage = renderer.create(<ViewImage flow={tflow} message={tflow.response}/>),
        tree = viewImage.toJSON()

    it('should render correctly', () => {
        expect(tree).toMatchSnapshot()
    })
})

describe('ViewServer Component', () => {
    let store = TStore(),
    setContentViewDescFn = jest.fn(),
    setContentFn = jest.fn()

    it('should render correctly and connect to state', () => {
        let provider = renderer.create(
            <Provider store={store}>
                <ViewServer
                    setContentViewDescription={setContentViewDescFn}
                    setContent={setContentFn}
                    flow={tflow}
                    message={tflow.response}
                />
            </Provider>),
            tree = provider.toJSON()
        expect(tree).toMatchSnapshot()

        let viewServer = renderer.create(
            <PureViewServer
                showFullContent={true}
                maxLines={10}
                setContentViewDescription={setContentViewDescFn}
                setContent={setContentViewDescFn}
                flow={tflow}
                message={tflow.response}
                content={JSON.stringify({lines: [['k1', 'v1']]})}
            />
        )
        tree = viewServer.toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('should handle componentWillReceiveProps', () => {
        // case of fail to parse content
        let viewSever = TestUtils.renderIntoDocument(
            <PureViewServer
                showFullContent={true}
                maxLines={10}
                setContentViewDescription={setContentViewDescFn}
                setContent={setContentViewDescFn}
                flow={tflow}
                message={tflow.response}
                content={JSON.stringify({lines: [['k1', 'v1']]})}
            />
        )
        viewSever.componentWillReceiveProps({...viewSever.props, content: '{foo' })
        let e = ''
        try {JSON.parse('{foo') } catch(err){ e = err.message}
        expect(viewSever.data).toEqual({ description: e, lines: [] })
    })
})

describe('Edit Component', () => {
    it('should render correctly', () => {
        let edit = renderer.create(<Edit
                content="foo"
                onChange={jest.fn}
                flow={tflow}
                message={tflow.response}
            />),
            tree = edit.toJSON()
        expect(tree).toMatchSnapshot()
    })
})
