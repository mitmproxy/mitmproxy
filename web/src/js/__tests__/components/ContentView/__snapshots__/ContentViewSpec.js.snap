// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Component should render correctly 1`] = `
<div
  className="text-center"
>
  <i
    className="fa fa-spinner fa-spin"
  />
</div>
`;

exports[`ViewImage Component should render correctly 1`] = `
<div
  className="flowview-image"
>
  <img
    alt="preview"
    className="img-thumbnail"
    src="./flows/d91165be-ca1f-4612-88a9-c0f8696f3e29/response/content.data"
  />
</div>
`;

exports[`ViewServer Component should render correctly and connect to state 1`] = `
<div
  className="text-center"
>
  <i
    className="fa fa-spinner fa-spin"
  />
</div>
`;

exports[`ViewServer Component should render correctly and connect to state 2`] = `
<div>
  <pre>
    <div>
      <span
        className="k"
      >
        1
      </span>
      <span
        className="v"
      >
        1
      </span>
    </div>
  </pre>
</div>
`;
